File: sketches/videotext/src/render/markup.cpp
// Tags are names, not colours: parsing is theme-independent, so one lint
// pass covers both themes. A repeated unknown tag warns only once, so a
// repeated typo doesn't spam the sender.

#include "markup.h"

namespace vt {
namespace {

// Indexed by Tag: a tag's id is the palette index of the colour it names.
constexpr const char32_t* TAG_NAMES[TAG_COUNT] = {U"black", U"white", U"green", U"blue", U"red", U"yellow", U"orange"};

Tag tagNamed(const Codepoints& word) {
    for (int i = 0; i < TAG_COUNT; i++) {
        if (word == TAG_NAMES[i]) return static_cast<Tag>(i);
    }
    return NO_TAG;
}

}  // namespace

Markup parseMarkup(const Codepoints& body) {
    Markup markup;
    std::vector<Codepoints> seenUnknown;
    Tag currentTag = NO_TAG;
    bool fenced = false;

    std::size_t lineStart = 0;
    while (true) {
        std::size_t lineEnd = body.find(U'\n', lineStart);
        if (lineEnd == Codepoints::npos) lineEnd = body.size();
        const Codepoints rawLine = body.substr(lineStart, lineEnd - lineStart);

        if (trimmed(rawLine) == U"```") {
            fenced = !fenced;  // fence marker lines are not rendered
        } else {
            MarkupLine line{fenced, {}};
            std::size_t i = 0;
            while (i < rawLine.size()) {
                if (rawLine[i] == U'{') {
                    const std::size_t end = rawLine.find(U'}', i);
                    if (end != Codepoints::npos) {
                        const Codepoints word = rawLine.substr(i + 1, end - i - 1);
                        if (word == U"/") {
                            currentTag = NO_TAG;
                            i = end + 1;
                            continue;
                        }
                        const Tag tag = tagNamed(word);
                        if (tag != NO_TAG) {
                            currentTag = tag;
                            i = end + 1;
                            continue;
                        }
                        const Codepoints literal = rawLine.substr(i, end - i + 1);
                        for (char32_t c : literal) line.cells.push_back({c, currentTag});
                        bool seen = false;
                        for (const Codepoints& s : seenUnknown) {
                            if (s == literal) seen = true;
                        }
                        if (!seen) {
                            seenUnknown.push_back(literal);
                            markup.warnings.push_back("unknown tag " + encodeUtf8(literal));
                        }
                        i = end + 1;
                        continue;
                    }
                }
                line.cells.push_back({rawLine[i], currentTag});
                i++;
            }
            markup.lines.push_back(std::move(line));
        }

        if (lineEnd == body.size()) break;
        lineStart = lineEnd + 1;
    }
    return markup;
}

}  // namespace vt

File: archive/relay/src/render/markup.js
// Body markup grammar: {red}{green}{blue}{yellow}{orange}{white}{black} set
// the active colour tag; {/} resets to the default. Unknown {word} tags are
// literal (rendered with the braces) and reported as a warning, deduplicated
// so a repeated typo doesn't spam the sender. Newlines are hard breaks.
// Fenced ``` blocks toggle a non-wrapping mode; tags still parse inside them.
// Theme-independent by design (see lint.js) — tags are names, not colours.

const COLOR_TAGS = new Set(["red", "green", "blue", "yellow", "orange", "white", "black"]);

export function parse(body) {
    const warnings = [];
    const seenUnknown = new Set();
    const lines = [];
    let currentTag = null;
    let fenced = false;
    let currentSpans = [];
    let currentText = "";

    const flushSpan = () => {
        if (currentText.length > 0) {
            currentSpans.push({ text: currentText, tag: currentTag });
            currentText = "";
        }
    };
    const flushLine = () => {
        flushSpan();
        lines.push({ fenced, spans: currentSpans });
        currentSpans = [];
    };

    for (const rawLine of body.split("\n")) {
        if (rawLine.trim() === "```") {
            fenced = !fenced;
            continue; // fence marker lines are not rendered
        }
        let i = 0;
        while (i < rawLine.length) {
            if (rawLine[i] === "{") {
                const end = rawLine.indexOf("}", i);
                if (end !== -1) {
                    const word = rawLine.slice(i + 1, end);
                    if (word === "/") {
                        flushSpan();
                        currentTag = null;
                        i = end + 1;
                        continue;
                    }
                    if (COLOR_TAGS.has(word)) {
                        flushSpan();
                        currentTag = word;
                        i = end + 1;
                        continue;
                    }
                    const literal = `{${word}}`;
                    currentText += literal;
                    if (!seenUnknown.has(literal)) {
                        seenUnknown.add(literal);
                        warnings.push(`unknown tag ${literal}`);
                    }
                    i = end + 1;
                    continue;
                }
            }
            currentText += rawLine[i];
            i++;
        }
        flushLine();
    }

    return { lines, warnings };
}
